### PR TITLE
[ip] Dropping `Ipv4Endpoint` and `Port16`

### PR DESCRIPTION
**Pull request by ppenna**
Description
-------------

This PR fixes https://github.com/demikernel/inetstack/issues/172.

Summary of Changes
------------------------
- Dropped `Port16` structure in favor of `u16`
- Dropped `Ipv4Endpoint` in favor of `std::net::SocketAddrV4`

Further Considerations
--------------------------
- https://github.com/demikernel/inetstack/issues/173 is related to this PR and shall be fixed next.
- https://github.com/demikernel/inetstack/issues/169 is related to this PR and shall be fixed next.

### CODE REVIEW

File: src/protocols/udp/datagram/header.rs [hunk 1]
```diff
@@ -42,9 +38,9 @@ pub const UDP_HEADER_SIZE: usize = 8;
 #[derive(Debug)]
 pub struct UdpHeader {
     /// Port used on sender side (optional).
-    src_port: Option<Port16>,
+    src_port: Option<u16>,
```
Comment (BrianZill) [inline src/protocols/udp/datagram/header.rs:41]: Get rid of the Option here (and everywhere else the src_port value is such); it is no longer needed.

File: src/protocols/tcp/isn_generator.rs [hunk 2]
```diff
@@ -26,21 +24,21 @@ impl IsnGenerator {
...
     }
 
     #[cfg(not(test))]
-    pub fn generate(&mut self, local: &Ipv4Endpoint, remote: &Ipv4Endpoint) -> SeqNumber {
+    pub fn generate(&mut self, local: &SocketAddrV4, remote: &SocketAddrV4) -> SeqNumber {
         let crc: crc::Crc<u32> = crc::Crc::<u32>::new(&crc::CRC_32_CKSUM);
         let mut digest = crc.digest();
-        let remote_addr: u32 = remote.get_address().into();
+        let remote_addr: u32 = u32::from_be_bytes(remote.ip().octets());
         digest.update(&remote_addr.to_be_bytes());
```
Comment (BrianZill) [inline src/protocols/tcp/isn_generator.rs:36]: This was already rather silly (why byte swap before hashing when the hash is just for randomization?), now it is extra silly (why byte swap to convert into a host endian u32 and then byte swap it right back to network endian for the digest call?).  Just feed digest.update() the address octets directly.

Same for local_addr below.

**Review by BrianZill — CHANGES_REQUESTED** [review]
Looks good except for two things:

- Remove the Option type for the UDP source port.  Treat it as a straight u16, just like the destination port.
- There are two instances of an extraneous conversion of address bytes to a u32 (including byte-swapping) and back again in the IsnGenerator code.  Please remove them.

File: src/protocols/tcp/isn_generator.rs (hunk 2 above)
Comment (ppenna) [reply src/protocols/tcp/isn_generator.rs:36]: Nice catch.

File: src/protocols/udp/datagram/header.rs (hunk 1 above)
Comment (ppenna) [reply src/protocols/udp/datagram/header.rs]: Thanks for pointing out this. Now we dropped unnecessary wrapping `Options` from IPv4 addresses as well.

File: src/protocols/udp/datagram/header.rs [hunk 3]
```diff
@@ -119,7 +119,7 @@ impl UdpHeader {
         let fixed_buf: &mut [u8; UDP_HEADER_SIZE] = (&mut buf[..UDP_HEADER_SIZE]).try_into().unwrap();
 
         // Write source port. If not present, write zeros.
```
Comment (BrianZill) [inline src/protocols/udp/datagram/header.rs:121]: If you have a moment to fix this comment (just delete the "If not present, write zeros."), please do.

**Review by BrianZill — APPROVED** [review]
This looks good now, I'm really happy we're getting rid of these non-standard types that complicated things without really improving anything.  Thanks for doing this.

The one thing I don't understand about this PR is the OperationResult::Pop change, probably because I'm still not entirely clear about how futures work.  I'll ask you about this later though and not hold up this PR.

File: src/protocols/udp/datagram/header.rs (hunk 3 above)
Comment (ppenna) [reply src/protocols/udp/datagram/header.rs:121]: Sure. I'll fix this before merging.

**Pull request merged**